User-created basemaps

annehaley opened the pull request:
This PR does the following:
- Create new model `Basemap` and a set of default objects that are created during migration
- Make client map creation/handling more abstract to allow switching basemap styles
- Add a menu to the controls bar for users to select a basemap
- Add a "New" button to the bottom of the basemaps menu which opens a dialog for creating a new basemaps
- In the basemap creation dialog, allow the user to specify the basemap with a tile URL or a style JSON

The following video demonstrates the new functionality:
https://drive.google.com/file/d/16pZCFmvi6-MuEhlnG6yqLWOy3wcJBSKl/view?usp=sharing

Resolves #240.

zachmullen commented on web/src/components/ControlsBar.vue:
```diff
@@ -11,6 +17,112 @@ const copyMenuShown = ref(false);
+  createBasemap({
+    name: newBasemapName.value,
+    style: newBasemapStyleJSON.value,
+  }).then((basemap) => {
```
Why old-school promise chaining here rather than `await`?

BryonLewis commented on web/src/components/ControlsBar.vue:
```diff
@@ -11,6 +17,112 @@ const copyMenuShown = ref(false);
+        const preview = new Map({
+          container: 'basemap-preview-' + basemap.id,
+          attributionControl: false,
+        });
```
Something I should improve in my toggle code as well is holding onto the map objects and calling `map.remove()` when done with them.  I don't know how well GC will grab and free up the resources and the elements from these models.

BryonLewis reviewed (changes requested):
I haven't had time yet to run branch in geoinsight.   I've only looked through the PR.
The only comment I feel is strongly reccomended is the one for guarding against adding bad styles into the list (testing and verification through the preview).  Other than that one all others can be marked as suggestions and feel free to push back on any of them.
Tomorrow I'll go a bit more into detail and test the branch but I don't think I would have much else.

BryonLewis commented on web/src/components/ControlsBar.vue:
```diff
@@ -11,6 +17,112 @@ const copyMenuShown = ref(false);
+      })
+    }, 10)
+  }
+}
```
I think it's fine as long as it works.  My worry was just that some client might have an issue, and the 10s timeout might not be enough to render the items and then display them.  I'm sure it would work in 99.9999% of cases, but figured we could avoid that chance if we had something that worked less based on time.

annehaley commented on web/src/components/ControlsBar.vue:
```diff
@@ -116,6 +257,54 @@ function takeScreenshot(save: boolean) {
+          <v-btn @click="cancelBasemapCreate" variant="tonal">
+            Cancel
+          </v-btn>
+          <v-btn color="primary" :disabled="!newBasemapName || !newBasemapStyleJSON" @click="submitBasemapCreate">
```
I found that the `on('error')` wasn't always firing, so I tried doing this validation before setting the style. In [996b12f](https://github.com/OpenGeoscience/geoinsight/pull/250/commits/996b12feb77d666249ddb48febdfe65c798aaeb8), I added a step to validate the style JSON using MapLibre's `validateStyleMin`. If any errors exist, the style won't be applied and the "create" button will be disabled.

annehaley commented on web/src/components/ControlsBar.vue:
```diff
@@ -116,6 +257,54 @@ function takeScreenshot(save: boolean) {
+          </v-btn>
+        </v-card-title>
+        <v-card-text class="d-flex" style="flex-direction: column; row-gap: 5px">
+          <v-text-field v-model="newBasemapName" label="Name" hide-details density="compact" autofocus />
```
The JSON editor has a debounce of 300ms by default (https://github.com/cloydlau/json-editor-vue?tab=readme-ov-file#props). We can increase that if it helps.

Also, with the changes in [996b12f](https://github.com/OpenGeoscience/geoinsight/pull/250/commits/996b12feb77d666249ddb48febdfe65c798aaeb8), the preview's style will only be set when the JSON contents are a valid style.

annehaley commented:
> The only other very minor suggestion is maybe moving the '+ New' from the bottom of the list to the heading or top where it has 'Base Map Options'. Just so you don't have to scroll to get to the adding new button.

Moved to the top of the basemaps list in [f77b94a](https://github.com/OpenGeoscience/geoinsight/pull/250/commits/f77b94a297f59505c29a602500b5aafed67c5659). I also updated the text and color to match the "+ Create Custom Colormap" button we have in the colormaps list
